### PR TITLE
#693: Add tests to JSON

### PR DESCRIPTION
**Pull request by bonkey**
#693: Add tests to JSON

### CODE REVIEW

**Comment by bonkey** [comment]
Tests for arrays and dictionaries still left, will add later on

**Comment by hani-ibrahim** [comment]
👍

**Comment by bonkey** [comment]
@hani-ibrahim 
Whoa, seems like real 80/20 rule... I got it covered all in the end. Not that everything was really needed, but it helped a lot to spot few inconsistencies and get a better interface. Something like Post-TDD.

**Comment by bonkey** [comment]
👍

**Comment by bonkey** [comment]
👍

**Comment by bonkey** [comment]
👍

File: AtlasSDK/AtlasSDK/JSON/JSON.swift [hunk 1]
```diff
@@ -54,28 +60,19 @@ extension JSON: CustomStringConvertible {
 
 extension JSON {
 
-    var arrayObject: [Any]? {
-        return self.rawObject as? [Any]
+    var jsons: [JSON] {
```
Comment (hani-ibrahim) [inline AtlasSDK/AtlasSDK/JSON/JSON.swift]: `jsonArray`?

File: AtlasSDK/AtlasSDKTests/JSON/JSONDictionariesTests.swift [hunk 2]
```diff
@@ -7,11 +7,9 @@ import Nimble
 
 @testable import AtlasSDK
 
-class JSONNumbersTests: JSONTestCase {
+class JSONDictionariesTests: JSONTestCase {
```
Comment (hani-ibrahim) [inline AtlasSDK/AtlasSDKTests/JSON/JSONDictionariesTests.swift]: unused class?

File: AtlasSDK/AtlasSDKTests/JSON/JSONNullsTests.swift [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    func testJSONFromNil() {
+        expect(JSON(nil)) == JSON.null
+    }
+
+    func testNullString() {
+        let nullString = self.json["nulls", "nullString"]
+        expect(nullString.string) == "null"
+        expect(nullString.bool).to(beNil())
+        expect(nullString.isNull) == false
+        expect(nullString != JSON.null) == false
```
Comment (hani-ibrahim) [inline AtlasSDK/AtlasSDKTests/JSON/JSONNullsTests.swift]: `nullString.isNull` is false while `nullString != JSON.null` is also false
so what is the difference between `isNull` and `null`?

**Comment by bonkey** [comment]
👍

File: AtlasSDK/AtlasSDK/JSON/JSON.swift (hunk 1 above)
Comment (bonkey) [inline AtlasSDK/AtlasSDK/JSON/JSON.swift]: Actually, I like it this way. I don't think it's needed to expose type in name, like `.articles`, not `.articleArray` or `.articlesArray`

File: AtlasSDK/AtlasSDKTests/JSON/JSONNullsTests.swift (hunk 3 above)
Comment (bonkey) [inline AtlasSDK/AtlasSDKTests/JSON/JSONNullsTests.swift]: l0hlusahkuyrxr77a

File: AtlasSDK/AtlasSDK/JSON/JSON.swift (hunk 1 above)
Comment (hani-ibrahim) [inline AtlasSDK/AtlasSDK/JSON/JSON.swift]: yeah I agree but `json.jsons` sounds bad 👹

File: AtlasSDK/AtlasSDKTests/JSON/JSONNullsTests.swift (hunk 3 above)
Comment (bonkey) [inline AtlasSDK/AtlasSDKTests/JSON/JSONNullsTests.swift]: @hani-ibrahim I was lying about nils... and got punished

bart-simpson-generator

**Comment by bonkey** [comment]
👍

File: AtlasSDK/AtlasSDK/JSON/JSON.swift [hunk 4]
```diff
@@ -121,12 +121,25 @@ extension JSON {
 extension JSON: Equatable { }
 
 func == (lhs: JSON, rhs: JSON) -> Bool {
+    if lhs.isNull && rhs.isNull {
+        return true
+    }
+    if let lhs = lhs.bool, let rhs = rhs.bool, lhs == rhs {
+        return true
```
Comment (hani-ibrahim) [inline AtlasSDK/AtlasSDK/JSON/JSON.swift]: why returning true and the `array` and `dictionary` cases return the the `==`

**Comment by hani-ibrahim** [comment]
👍

File: AtlasSDK/AtlasSDK/JSON/JSON.swift (hunk 4 above)
Comment (bonkey) [inline AtlasSDK/AtlasSDK/JSON/JSON.swift]: Ah, of course!

Comment (bonkey) [inline AtlasSDK/AtlasSDK/JSON/JSON.swift]: Result is the same, but it's more readable.

**Comment by hani-ibrahim** [comment]
👍

**Comment by bonkey** [comment]
👍

**Pull request merged**